src/link/script.y

```diff
@@ -0,0 +1,479 @@
+			               strCaseCmp)) {
+				return keyword.tokenGen();
+			}
+		}
+
+		scriptError(context, "Unknown keyword \"%s\"", ident.c_str());
+		goto try_again; // Try lexing another token.
+	} else {
+		scriptError(context, "Unexpected character '%s'", printChar(c));
+		// Keep reading characters until the EOL, to avoid reporting too many errors.
+		for (c = context.file.sgetc(); !isNewline(c); c = context.file.sgetc()) {
+			if (c == LexerStackEntry::eof) {
+				break;
+			}
+		}
+		goto try_again;
+	}
+	// Not marking as unreachable; this will generate a warning if any codepath forgets to return.
+}
+
+// Semantic actions.
+
+static SectionType activeType;
+static uint32_t activeBankIdx; // This is the index into the array, not
+static std::array<std::vector<uint16_t>, SECTTYPE_INVALID> curAddr;
+
+static void setActiveTypeAndIdx(SectionType type, uint32_t idx) {
+	activeType = type;
+	activeBankIdx = idx;
+	if (curAddr[activeType].size() <= activeBankIdx) {
+		curAddr[activeType].resize(activeBankIdx + 1, sectionTypeInfo[type].startAddr);
+	}
+}
+
+static void setSectionType(SectionType type) {
+	auto const &context = lexerStack.back();
+
+	if (nbbanks(type) != 1) {
+		scriptError(context, "A bank number must be specified for %s",
+		            sectionTypeInfo[type].name.c_str());
+		// Keep going with a default value for the bank index.
+	}
+
+	setActiveTypeAndIdx(type, 0); // There is only a single bank anyway, so just set the index to 0.
+}
+
+static void setSectionType(SectionType type, uint32_t bank) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[type];
+
+	if (bank < typeInfo.firstBank) {
+		scriptError(context, "%s bank %" PRIu32 " doesn't exist, the minimum is %" PRIu32,
+		            typeInfo.name.c_str(), bank, typeInfo.firstBank);
+		bank = typeInfo.firstBank;
+	} else if (bank > typeInfo.lastBank) {
+		scriptError(context, "%s bank %" PRIu32 " doesn't exist, the maximum is %" PRIu32,
+		            typeInfo.name.c_str(), bank, typeInfo.lastBank);
+	}
+
+	setActiveTypeAndIdx(type, bank - typeInfo.firstBank);
+}
+
+static void setAddr(uint32_t addr) {
+	auto const &context = lexerStack.back();
+	auto &pc = curAddr[activeType][activeBankIdx];
+	auto const &typeInfo = sectionTypeInfo[activeType];
+
+	if (addr < pc) {
+		scriptError(context, "ORG cannot be used to go backwards (from $%04x to $%04x)", pc, addr);
+	} else if (addr > endaddr(activeType)) { // Allow "one past the end" sections.
+		scriptError(context, "Cannot go to $%04" PRIx32 ": %s ends at $%04" PRIx16 "",
+		            addr, typeInfo.name.c_str(), endaddr(activeType));
+		pc = endaddr(activeType);
+	} else {
+		pc = addr;
+	}
+}
+
+static void alignTo(uint32_t alignment, uint32_t alignOfs) {
+	auto const &context = lexerStack.back();
+	auto const &typeInfo = sectionTypeInfo[activeType];
+	auto &pc = curAddr[activeType][activeBankIdx];
+
+	// TODO: maybe warn if truncating?
```

Comment:
Yes, RGBASM does.

Reply:
Yeah, the problem was that I wasn't feeling like adding warning infrastructure to RGBLINK, and the old parser wasn't warning anyway, so I didn't bother adding the complexity right now.

Reply:
Fair. And if it wasn't an error in the previous script.cpp, no need to make it one now.

Reply:
Filed as #1268.